File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int lo = 0;
        int hi = 0;
        for(int w : weights){
            lo = Math.max(lo, w);
            hi += w;
        }
        int ans = hi;
        while(lo<=hi){
            int mid = lo + (hi-lo)/2;
            if(check(weights, days, mid)){
                ans = mid;
                hi = mid -1;
            }
            else{
                lo = mid + 1;
            }
        }
        return ans;
    }
    
    public boolean check(int[] weights, int days, int mid){
        int sum = 0;
        int count = 0;
        for(int w: weights){
            sum += w;
            if(sum > mid){
                count++;
                sum = w;
                if(count == days){
                    return false;
                }
            }
        }
        return true;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def shipWithinDays(self, weights, days):

        l=max(weights)
        r=sum(weights)
        res=r

        def check(cap):
            ships=1
            currentcapacity=cap
            for i in weights:
                if currentcapacity-i<0:
                    ships+=1
                    currentcapacity=cap
                currentcapacity-=i
            return ships<=days

        while l<=r:
            mid=l+((r-l)//2)
            if check(mid):
                res=min(res,mid)
                r=mid-1
            else:
                l=mid+1
        return res